server/src/test/java/org/elasticsearch/action/bulk/BulkProcessorTests.java

```diff
@@ -90,6 +105,180 @@ public void testBulkProcessorFlushPreservesContext() throws InterruptedException
+        CountDownLatch startGate = new CountDownLatch(1);
+
+        IndexRequest indexRequest = new IndexRequest();
+        String bulkRequest = "{ \"index\" : { \"_index\" : \"test\", \"_id\" : \"1\" } }\n" +  "{ \"field1\" : \"value1\" }\n";
+        BytesReference bytesReference =
+            BytesReference.fromByteBuffers(new ByteBuffer[]{ByteBuffer.wrap(bulkRequest.getBytes(StandardCharsets.UTF_8))});
+        List<Future> futures = new ArrayList<>();
+        for (final AtomicInteger i = new AtomicInteger(0); i.getAndIncrement() < maxDocuments; ) {
+            futures.add(executorService.submit(() -> {
+                try {
+                    //don't start any work until all tasks are submitted
+                    startGate.await();
+                    //alternate between ways to add to the bulk processor
+                    if (i.get() % 2 == 0) {
+                        bulkProcessor.add(indexRequest);
+                    } else {
+                        bulkProcessor.add(bytesReference, null, null, XContentType.JSON);
+                    }
+                } catch (Exception e) {
+                    throw new RuntimeException(e);
+                }
+            }));
+        }
+        startGate.countDown();
+
+        for (Future f : futures) {
+            try {
+                f.get();
+            } catch (Exception e){
+                failureCount.incrementAndGet();
+                logger.error("failure while getting future", e);
+            }
+        }
+        executorService.shutdown();
+        executorService.awaitTermination(10, TimeUnit.SECONDS);
+
+        if (failureCount.get() > 0 || successCount.get() != expectedExecutions || requestCount.get() != successCount.get()) {
+            fail("\nExpected Bulks: " + expectedExecutions + "\n" +
+                "Requested Bulks: " + requestCount.get() + "\n" +
+                "Successful Bulks: " + successCount.get() + "\n" +
+                "Failed Bulks: " + failureCount.get() + "\n" +
+                "Max Documents: " + maxDocuments + "\n" +
+                "Max Batch Size: " + maxBatchSize + "\n" +
+                "Concurrent Clients: " + concurrentClients + "\n" +
+                "Concurrent Bulk Requests: " + concurrentBulkRequests + "\n"
+            );
+        }
+        bulkProcessor.close();
+        //count total docs after processor is closed since there may have been partial batches that are flushed on close.
+        assertEquals(docCount.get(), maxDocuments);
+    }
+
+    public void testConcurrentExecutionsWithFlush() throws Exception {
+        final int maxDocuments = 100_000;
+        final int concurrentClients = 2;
+        final int maxBatchSize = Integer.MAX_VALUE; //don't flush based on size
+        final int concurrentBulkRequests = randomIntBetween(0, 20);
+        final int simulateWorkTimeInMillis = 5;
+        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[]{new BulkItemResponse()}, 0);
```

Comment:
this is just my preference, feel free to ignore:

```suggestion
        BulkResponse bulkResponse = new BulkResponse(new BulkItemResponse[] { new BulkItemResponse() }, 0);
```